### DISCUSSION TITLE
Tracking the Route path for logging purposes

### DISCUSSION THREAD

**Discussion by danielearwicker** [remix-run/remix#7472] [Q&A]
I have an integration with a logging platform that logs the location path of my app as the user navigates around it.

But sadly this isn't that useful because the routes include dynamic segments like `:productId` which are filled in with actual values, and this is just noise when the purposes of these logs is to build up a picture of how often users look at different areas of our site. As the location's path ends up being `/main/products/723876/details`, each view of that page has a different `productId` embedded in it and so I can't easily count all the visits to that page.

But our `Route` tree gives static placeholder names (such as `:productId`) to those slots. If I could track whenever the route changes and get a path that still had the static placeholder names in it, e.g. `/main/products/:productId/details` that would work great as a kind of key for stats.

Is there any way for me to get a notification when the location changes and recover this non-substituted path for the new location?

**Comment by kiliman** [comment]
Yes, Remix "kind of" provides all the tools to do this. First, you want to use the `matchRoutes`[^1] helper from React Router. This takes the path, like `/main/products/723876/details`, and returns all the matching route configs. You'll typically want the leaf route config.

However, this helper requires the routes to be configured the way React Router expects for `createBrowserRouter`[^2].

Luckily Remix has a function that will cover the Remix route manifest to the one RR needs. `createStaticDataRoutes`[^3]

Most likely you'll want to pass the `ServerBuild` manifest in `getLoadContext` of your server entry (if you're using Remix App Server, you'll need to convert to Express).

I've got a semi-working example. When I get a chance to clean it up, I'll post it. At least this will point you in the right direction.

[^1]: https://reactrouter.com/en/main/utils/match-routes
[^2]: https://reactrouter.com/en/main/routers/create-browser-router
[^3]: https://github.com/remix-run/remix/blob/7f4168f699581a978925e167d804bf3bfa639cf5/packages/remix-server-runtime/routes.ts#L73
